tools/lsp/main.rs

```diff
@@ -33,31 +33,112 @@ use std::task::{Poll, Waker};
 struct Previewer {
     #[allow(unused)]
     server_notifier: ServerNotifier,
+    use_external_previewer: RefCell<bool>,
 }
 
 impl PreviewApi for Previewer {
+    fn set_use_external_previewer(&self, ctx: &Rc<Context>, use_external: bool) {
+        let current_is_external = {
+            let mut tmp = self.use_external_previewer.borrow_mut();
+            let old = *tmp;
+            *tmp = use_external;
+            old
+        };
+
+        if current_is_external != use_external {
+            let documents = &ctx.document_cache.borrow().documents;
+
+            for (p, d) in documents.all_file_documents() {
+                let Some(node) = &d.node else {
+                    continue;
+                };
+                self.set_contents(p, &node.text().to_string());
+            }
+            let style = documents.compiler_config.style.clone().unwrap_or_default();
+            self.config_changed(&style, &documents.compiler_config.include_paths);
+
+            if use_external {
+                preview::close_ui();
+            }
+        }
+    }
+
     fn set_contents(&self, _path: &std::path::Path, _contents: &str) {
         #[cfg(feature = "preview")]
```

Comment:
So this preview feature is both for the internal AND external previewer?
I'm wondering if it wouldn't be beneficial to keep that as two features

Reply:
ok it's fine as long as it's documented in cargo.toml what the feature are for.